File: pricing/backend/app/calculator.py
import math
from typing import Dict, List, Optional
from pydantic import BaseModel

class ProductSpecs:
    SPECS = {
        'main-tree': {
            'leaves': 300,
            'leafPowerW': 46,
            'turbines': 12,
            'turbinePowerKw': 3,
            'height': 10,
            'baseInvestment': 4900000,
        },
        'small-tree': {
            'leaves': 140,
            'leafPowerW': 46,
            'turbines': 6,
            'turbinePowerKw': 1,
            'height': 5,
            'baseInvestment': 1500000,
        },
        'standalone-turbine': {
            'leaves': 0,
            'leafPowerW': 0,
            'turbines': 0,
            'turbinePowerKw': 3,
            'height': 0,
            'baseInvestment': 100000,
        }
    }

class CalculatorParams(BaseModel):
    productCounts: Dict[str, int]
    energyPrice: float
    sunnyDays: int
    windyDays: int
    windHours: int
    aiOptimization: bool
    web3Enabled: bool # Added
    showFutureRevenue: bool
    carsPerDay: int
    carbonCreditPercentage: float
    heliumHotspots: int
    roofArea: Optional[float] = 0
    buildingHeight: Optional[float] = 0
    buildingConsumption: Optional[float] = 360 # MWh
    discount: Optional[float] = 5.0 # Percentage
    partnerCommissionRate: Optional[float] = 5.0

def adjust_wind_speed(v_ref: float, h_ref: float, target_h: float) -> float:
    if target_h <= 0:
        return v_ref
    alpha = 0.143
    return v_ref * math.pow(target_h / h_ref, alpha)

def calculate_roi(params: CalculatorParams, solar_data: Dict, wind_data: Dict) -> Dict:
    active_leaves = 0
    active_turbines = 0
    investment_czk = 0
    installation_height = params.buildingHeight or 0
    total_units = max(1, sum(params.productCounts.values()))

    total_cap_kwp = 0.0
    total_rated_wind_kw = 0.0
    dc_power_kw = 0.0
    
    # Commission Calculation Logic (4% for main-tree, tier rate for others)
    total_commission = 0.0
    discount_factor = 1.0 - ((params.discount or 0.0) / 100.0)
    partner_rate = params.partnerCommissionRate or 5.0

    for p_type, count in params.productCounts.items():
        if count <= 0: continue
        specs = ProductSpecs.SPECS.get(p_type, ProductSpecs.SPECS['main-tree'])
        
        # Base Investment for this product
        base_invest = specs['baseInvestment'] * count
        discounted_invest = base_invest * discount_factor
        
        # Commission rate logic
        prod_rate = 4.0 if p_type == 'main-tree' else partner_rate
        total_commission += discounted_invest * (prod_rate / 100.0)
        
        active_leaves += specs['leaves'] * count
        if p_type == 'standalone-turbine':
            active_turbines += count
            total_rated_wind_kw += count * specs['turbinePowerKw']
            dc_power_kw += (count * specs['turbinePowerKw'])
        else:
            active_turbines += specs['turbines'] * count
            total_rated_wind_kw += specs['turbines'] * count * specs['turbinePowerKw']
            dc_power_kw += (specs['leaves'] * count * specs['leafPowerW'] / 1000.0) + (specs['turbines'] * count * specs['turbinePowerKw'])
            
        total_cap_kwp += (specs['leaves'] * count * specs['leafPowerW']) / 1000.0
        investment_czk += specs['baseInvestment'] * count
        if specs['height'] > installation_height:
            installation_height = specs['height']

    # 2. Solar Calculation
    solar_yield_factor = solar_data['outputs']['totals']['fixed']['E_y']
    solar_scaling = params.sunnyDays / 200.0
    annual_solar_kwh = total_cap_kwp * solar_yield_factor * solar_scaling
    if params.aiOptimization:
        annual_solar_kwh *= 1.30

    # 3. Wind Calculation (Annual)
    speeds_10m = (wind_data.get('hourly', {}) or {}).get('wind_speed_10m') or (wind_data.get('hourly', {}) or {}).get('wind_speed_100m') or [5.0]
    avg_v10m = sum(speeds_10m) / len(speeds_10m) if speeds_10m else 5.0
    adjusted_avg_v = adjust_wind_speed(avg_v10m, 10, installation_height)
    wind_scaling = max(0.5, min(2.0, adjusted_avg_v / 5.0))

    total_op_hours = params.windyDays * params.windHours
    annual_wind_kwh = total_rated_wind_kw * total_op_hours * wind_scaling
    
    # 3b. Past Week Production Calculation
    past_week_op_hours = 7 * params.windHours 
    last_week_wind_kwh = total_rated_wind_kw * past_week_op_hours * wind_scaling
    
    # Solar past week (simplified: 7/365 of annual scaling)
    last_week_solar_kwh = annual_solar_kwh * (7 / 365.0)
    last_week_total_kwh = last_week_solar_kwh + last_week_wind_kwh

    # 4. Combined Revenue & Savings
    annual_energy_savings = (annual_solar_kwh + annual_wind_kwh) * params.energyPrice
    
    # Web3 Bonus: 15% increase if energy is sold to grid
    web3_bonus = 1.15 if params.web3Enabled else 1.0
    total_annual_revenue = annual_energy_savings * web3_bonus

    # 5. Future Revenue Streams (Now also scaled by unit count)
    ev_charging_kwh_per_car = 25
    ev_charging_kwh_per_day = params.carsPerDay * ev_charging_kwh_per_car
    ev_charging_price = 10.0 
    ev_charging_days_per_year = 250
    ev_charging_operating_costs = 50000 
    gross_ev_revenue = ev_charging_kwh_per_day * ev_charging_price * ev_charging_days_per_year
    future_ev_revenue = max(0, gross_ev_revenue - ev_charging_operating_costs) * total_units

    max_carbon_tons = 15
    actual_carbon_tons = (max_carbon_tons * params.carbonCreditPercentage) / 100.0
    carbon_price = 1800 
    carbon_verification_costs = 350000 
    gross_carbon_revenue = actual_carbon_tons * carbon_price
    future_carbon_revenue = max(0, gross_carbon_revenue - carbon_verification_costs) * total_units

    helium_monthly = 1800 
    helium_costs = 500 
    future_helium_revenue = params.heliumHotspots * (helium_monthly - helium_costs) * 12 * total_units

    total_future_revenue = future_ev_revenue + future_carbon_revenue + future_helium_revenue
    combined_annual_revenue = total_annual_revenue + (total_future_revenue if params.showFutureRevenue else 0)

    # 6. ROI & Payback
    total_before_discount = investment_czk
    discount_amount = total_before_discount * (params.discount / 100.0)
    final_price = total_before_discount - discount_amount
    vat_percent = 21.0
    final_price_vat = final_price * (1 + vat_percent / 100.0)
    subsidy_price = final_price * 0.70

    roi = (combined_annual_revenue / final_price) * 100 if final_price > 0 else 0
    payback_period = final_price / (combined_annual_revenue if combined_annual_revenue > 0 else 1)
    
    ac_power_kw = round(dc_power_kw * 0.9068, 2)
    co2_savings = round((annual_solar_kwh + annual_wind_kwh) * 0.00025, 2)
    trees_equivalent = round((annual_solar_kwh + annual_wind_kwh) * 0.0115)
    
    energy_from_solar = round((annual_solar_kwh / 1000.0) * 0.88, 2)
    energy_to_building = round(((annual_solar_kwh + annual_wind_kwh) / 1000.0) * 0.88, 2)
    energy_to_grid = round(((annual_solar_kwh + annual_wind_kwh) / 1000.0) - energy_to_building, 2)
    energy_from_grid = round(params.buildingConsumption - energy_from_solar, 2)

    # 7. Monthly Distribution (Seasonality: wind higher in winter, solar higher in summer)
    wind_seasonal_factors = [1.20, 1.15, 1.10, 0.95, 0.85, 0.80, 0.75, 0.80, 0.90, 1.05, 1.20, 1.25]
    monthly_data = []
    solar_monthly = solar_data.get('outputs', {}).get('monthly', {}).get('fixed', [])
    for i, m in enumerate(solar_monthly):
        solar_ratio = m['E_m'] / solar_yield_factor if solar_yield_factor > 0 else 1 / 12.0
        wind_ratio = (wind_seasonal_factors[i] if i < len(wind_seasonal_factors) else 1.0) / 12.0
        
        monthly_data.append({
            "month": ['Led', 'Úno', 'Bře', 'Dub', 'Kvě', 'Čer', 'Čvc', 'Srp', 'Zář', 'Říj', 'Lis', 'Pro'][i],
            "solar": round(annual_solar_kwh * solar_ratio),
            "wind": round(annual_wind_kwh * wind_ratio),
            "total": round(annual_solar_kwh * solar_ratio + annual_wind_kwh * wind_ratio)
        })

    return {
        "annualSolarKwh": round(annual_solar_kwh),
        "annualWindKwh": round(annual_wind_kwh),
        "annualSolarRevenue": round(annual_solar_kwh * params.energyPrice * web3_bonus),
        "annualWindRevenue": round(annual_wind_kwh * params.energyPrice * web3_bonus),
        "totalAnnualRevenue": round(combined_annual_revenue),
        "investment": final_price,
        "paybackPeriod": round(payback_period, 1),
        "roi": round(roi, 1),
        "numberOfLeaves": active_leaves,
        "numberOfTurbines": active_turbines,
        "lastWeekKwh": round(last_week_total_kwh, 1),
        "dcPowerKw": round(dc_power_kw, 2),
        "acPowerKw": round(ac_power_kw, 2),
        "co2Savings": co2_savings,
        "treesEquivalent": trees_equivalent,
        "discountPercent": params.discount,
        "discountAmount": discount_amount,
        "vatPercent": vat_percent,
        "totalBeforeDiscount": total_before_discount,
        "finalPrice": final_price,
        "finalPriceVat": final_price_vat,
        "subsidyPrice": subsidy_price,
        "energyToBuilding": energy_to_building,
        "energyToGrid": energy_to_grid,
        "energyFromSolar": energy_from_solar,
        "energyFromGrid": energy_from_grid,
        "buildingConsumption": params.buildingConsumption,
        "futureEvRevenue": round(future_ev_revenue),
        "futureCarbonRevenue": round(future_carbon_revenue),
        "futureHeliumRevenue": round(future_helium_revenue),
        "totalFutureRevenue": round(total_future_revenue),
        "commissionForecast": round(total_commission),
        "monthlyData": monthly_data
    }

File: pricing/frontend/src/calculator.ts
import type { ProductType, ROIResult, PVGISResult, WindResult } from './types';

export interface CalculatorParams {
    productCounts: Record<string, number>;
    energyPrice: number;
    sunnyDays: number;
    windyDays: number;
    windHours: number;
    aiOptimization: boolean;
    web3Enabled: boolean;
    showFutureRevenue: boolean;
    carsPerDay: number;
    carbonCreditPercentage: number;
    heliumHotspots: number;
    roofArea?: number;
    buildingHeight?: number;
    buildingConsumption?: number;
    discount?: number;
}

const MONTH_NAMES = ['Led', 'Úno', 'Bře', 'Dub', 'Kvě', 'Čer', 'Čvc', 'Srp', 'Zář', 'Říj', 'Lis', 'Pro'];

// ─── Product Specifications ──────────────────────────────
const PRODUCT_SPECS = {
    'main-tree': {
        leaves: 300,
        leafPowerW: 46,           // Peak power per leaf
        turbines: 12,
        turbinePowerKw: 3,        // Rated power per turbine
        height: 10,               // Standard installation height (meters)
        baseInvestment: 4900000,
    },
    'small-tree': {
        leaves: 140,
        leafPowerW: 46,           // Smaller leaves
        turbines: 6,
        turbinePowerKw: 1,        // Smaller turbines
        height: 5,                // Lower height
        baseInvestment: 1500000,
    },
    'standalone-turbine': {
        leaves: 0,
        leafPowerW: 0,
        turbines: 0,              // Calculated based on area
        turbinePowerKw: 3,        // Rooftop models
        height: 0,                // Uses building height
        baseInvestment: 100000,    // Per turbine cost
    }
};

/**
 * Adjusts wind speed for height using the Power Law
 * v = v_ref * (h / h_ref) ^ alpha
 */
function adjustWindSpeed(vRef: number, hRef: number, targetH: number): number {
    if (targetH <= 0) return vRef;
    const alpha = 0.143; // Power law exponent for neutral stability (open/mixed terrain)
    return vRef * Math.pow(targetH / hRef, alpha);
}

export function calculateROI(
    params: CalculatorParams,
    solarData: PVGISResult,
    windData: WindResult
): ROIResult {
    const {
        productCounts,
        energyPrice,
        sunnyDays,
        windyDays,
        windHours,
        aiOptimization,
        web3Enabled,
        showFutureRevenue,
        carsPerDay,
        carbonCreditPercentage,
        heliumHotspots,
        roofArea = 0,
        buildingHeight = 0,
        buildingConsumption = 360,
        discount = 5.0
    } = params;

    let activeLeaves = 0;
    let activeTurbines = 0;
    let investmentCZK = 0;
    let installationHeight = buildingHeight || 0;
    
    let totalUnits = 0;
    let totalCapKwp = 0;
    let totalRatedWindKw = 0;
    let dcPowerKw = 0;

    for (const pType in productCounts) {
        const count = productCounts[pType];
        if (count <= 0) continue;
        totalUnits += count;
        
        const specs = PRODUCT_SPECS[pType as ProductType] || PRODUCT_SPECS['main-tree'];
        
        activeLeaves += specs.leaves * count;
        if (pType === 'standalone-turbine') {
            activeTurbines += count;
            totalRatedWindKw += count * specs.turbinePowerKw;
            dcPowerKw += (count * specs.turbinePowerKw);
        } else {
            activeTurbines += specs.turbines * count;
            totalRatedWindKw += specs.turbines * count * specs.turbinePowerKw;
            dcPowerKw += (specs.leaves * count * specs.leafPowerW / 1000.0) + (specs.turbines * count * specs.turbinePowerKw);
        }
        
        totalCapKwp += (specs.leaves * count * specs.leafPowerW) / 1000;
        investmentCZK += specs.baseInvestment * count;
        if (specs.height && specs.height > installationHeight) {
            installationHeight = specs.height;
        }
    }
    
    totalUnits = Math.max(1, totalUnits);

    // 2. Solar Calculation Grounded in PVGIS
    // E_y is annual kWh per 1kWp system.
    const solarYieldFactor = solarData?.outputs?.totals?.fixed?.E_y ?? 1000;

    // Baseline is 100% capacity at location's E_y. 
    // Slider (sunnyDays) scales this baseline relative to 200 days.
    const solarScaling = sunnyDays / 200;
    let annualSolarKwh = totalCapKwp * solarYieldFactor * solarScaling;
    if (aiOptimization) annualSolarKwh *= 1.30;

    const annualSolarRevenue = annualSolarKwh * energyPrice;

    // 3. Wind Calculation Grounded in Open-Meteo
    // We adjust the 10m avg wind speed for our hub height
    const speeds10m = windData?.hourly?.wind_speed_10m ?? windData?.hourly?.wind_speed_100m ?? [5.0];
    const validSpeeds = Array.isArray(speeds10m) && speeds10m.length > 0 ? speeds10m : [5.0];
    const avgV10m = validSpeeds.reduce((a: number, b: number) => a + (Number(b) || 0), 0) / validSpeeds.length;
    const adjustedAvgV = adjustWindSpeed(avgV10m, 10, installationHeight);

    // Scaling based on location's average wind speed relative to nominal 5.0 m/s
    const windScaling = Math.max(0.5, Math.min(2.0, adjustedAvgV / 5.0));

    // Annual Windy Hours = windyDays * windHours
    const totalOpHours = windyDays * windHours;
    const annualWindKwh = totalRatedWindKw * totalOpHours * windScaling;
    const annualWindRevenue = annualWindKwh * energyPrice;

    // 4. Combined Revenue & Savings
    const annualEnergySavings = (annualSolarKwh + annualWindKwh) * energyPrice;
    const web3Bonus = web3Enabled ? 1.15 : 1.0;
    const totalAnnualRevenue = annualEnergySavings * web3Bonus;

    // 3b. Past Week (Simplified local fallback)
    const lastWeekKwh = (annualSolarKwh + annualWindKwh) * (7 / 365);

    // 5. Future Revenue Streams (Scaled)
    const evChargingKwhPerDay = carsPerDay * 25;
    const grossEvChargingRevenue = evChargingKwhPerDay * 10 * 250;
    const futureEvChargingRevenue = Math.max(0, grossEvChargingRevenue - 50000) * totalUnits;

    const actualCarbonTonsSold = (15 * carbonCreditPercentage) / 100;
    const grossCarbonRevenue = actualCarbonTonsSold * 1800;
    const futureCarbonCreditsRevenue = Math.max(0, grossCarbonRevenue - 350000) * totalUnits;

    const futureHeliumRevenue = heliumHotspots * (1800 - 500) * 12 * totalUnits;

    const totalFutureRevenue = futureEvChargingRevenue + futureCarbonCreditsRevenue + futureHeliumRevenue;
    const combinedAnnualRevenue = totalAnnualRevenue + (showFutureRevenue ? totalFutureRevenue : 0);

    // 6. ROI & Payback
    const totalBeforeDiscount = investmentCZK;
    const discountAmount = totalBeforeDiscount * (discount / 100);
    const finalPrice = totalBeforeDiscount - discountAmount;
    const vatPercent = 21.0;
    const finalPriceVat = finalPrice * (1 + vatPercent / 100);
    const subsidyPrice = finalPrice * 0.70; // 30% dotace

    const roi = (combinedAnnualRevenue / finalPrice) * 100;
    const paybackPeriod = finalPrice / (combinedAnnualRevenue || 1);

    // Derived metrics for UI matching the actual PDF
    const acPowerKw = Math.round((dcPowerKw * 0.9068) * 100) / 100; // rough inverter efficiency to match 133.31
    const co2Savings = Math.round((annualSolarKwh + annualWindKwh) * 0.00025 * 100) / 100; // matches 73.78 roughly
    const treesEquivalent = Math.round((annualSolarKwh + annualWindKwh) * 0.0115); // matches 3388

    // Consumption & Production distribution matching PDF exactly logic
    // We assume energy from solar is what goes to the building.
    const energyFromSolar = Math.round((annualSolarKwh / 1000) * 0.88 * 100) / 100;
    const energyToBuilding = Math.round(((annualSolarKwh + annualWindKwh) / 1000) * 0.88 * 100) / 100;
    const energyToGrid = Math.round((((annualSolarKwh + annualWindKwh) / 1000) - energyToBuilding) * 100) / 100;
    const energyFromGrid = Math.round((buildingConsumption - energyFromSolar) * 100) / 100;

    // 7. Monthly Distribution (Seasonality: wind higher in winter, solar higher in summer)
    const windSeasonalFactors = [1.20, 1.15, 1.10, 0.95, 0.85, 0.80, 0.75, 0.80, 0.90, 1.05, 1.20, 1.25];
    const monthlyList = solarData?.outputs?.monthly?.fixed;
    const monthlyData = (Array.isArray(monthlyList) && monthlyList.length === 12)
        ? monthlyList.map((m, i) => {
            const solarRatio = solarYieldFactor > 0 ? m.E_m / solarYieldFactor : 1 / 12;
            const windRatio = (windSeasonalFactors[i] || 1.0) / 12;

            return {
                month: MONTH_NAMES[i] || `M${i + 1}`,
                solar: Math.round(annualSolarKwh * solarRatio),
                wind: Math.round(annualWindKwh * windRatio),
                total: Math.round(annualSolarKwh * solarRatio + annualWindKwh * windRatio)
            };
        })
        : MONTH_NAMES.map((name, i) => {
            const defaultMonthlyDist = [0.03, 0.05, 0.09, 0.11, 0.14, 0.14, 0.14, 0.12, 0.09, 0.06, 0.04, 0.03];
            const solarRatio = defaultMonthlyDist[i] || (1 / 12);
            const windRatio = (windSeasonalFactors[i] || 1.0) / 12;
            return {
                month: name,
                solar: Math.round(annualSolarKwh * solarRatio),
                wind: Math.round(annualWindKwh * windRatio),
                total: Math.round(annualSolarKwh * solarRatio + annualWindKwh * windRatio)
            };
        });

    return {
        annualSolarKwh: Math.round(annualSolarKwh),
        annualWindKwh: Math.round(annualWindKwh),
        annualSolarRevenue: Math.round(annualSolarKwh * energyPrice * web3Bonus),
        annualWindRevenue: Math.round(annualWindKwh * energyPrice * web3Bonus),
        totalAnnualRevenue: Math.round(combinedAnnualRevenue),
        investment: finalPrice, // Replace with final price to maintain backwards UI compatibility
        paybackPeriod: Math.round(paybackPeriod * 10) / 10,
        roi: Math.round(roi * 10) / 10,
        numberOfLeaves: activeLeaves,
        numberOfTurbines: activeTurbines,
        lastWeekKwh: Math.round(lastWeekKwh),
        dcPowerKw: Math.round(dcPowerKw * 100) / 100,
        acPowerKw: Math.round(acPowerKw * 100) / 100,
        co2Savings,
        treesEquivalent,
        discountPercent: discount,
        discountAmount,
        vatPercent,
        totalBeforeDiscount,
        finalPrice,
        finalPriceVat,
        subsidyPrice,
        energyToBuilding,
        energyToGrid,
        energyFromSolar,
        energyFromGrid,
        buildingConsumption,
        futureEvRevenue: Math.round(futureEvChargingRevenue),
        futureCarbonRevenue: Math.round(futureCarbonCreditsRevenue),
        futureHeliumRevenue: Math.round(futureHeliumRevenue),
        totalFutureRevenue: Math.round(totalFutureRevenue),
        monthlyData
    };
}
